### PR TITLE
[Backport 23.11] yazi: shell integration

### PR DESCRIPTION
**Pull request by XYenon**
### Description

Backport https://github.com/nix-community/home-manager/pull/4699 https://github.com/nix-community/home-manager/pull/4746 https://github.com/nix-community/home-manager/pull/4846 to 23.11

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: tests/modules/programs/yazi/bash-integration-enabled.nix [hunk 1]
```diff
@@ -4,7 +4,7 @@ let
   shellIntegration = ''
     function ya() {
       tmp="$(mktemp -t "yazi-cwd.XXXXX")"
```
Comment (sxyazi) [inline tests/modules/programs/yazi/bash-integration-enabled.nix:6]:
```suggestion
      local tmp="$(mktemp -t "yazi-cwd.XXXXX")"
```

File: tests/modules/programs/yazi/zsh-integration-enabled.nix [hunk 2]
```diff
@@ -4,7 +4,7 @@ let
   shellIntegration = ''
     function ya() {
       tmp="$(mktemp -t "yazi-cwd.XXXXX")"
-      yazi --cwd-file="$tmp"
+      yazi "$@" --cwd-file="$tmp"
```
Comment (sxyazi) [inline tests/modules/programs/yazi/zsh-integration-enabled.nix:7]:
```suggestion
      local yazi "$@" --cwd-file="$tmp"
```

**Comment by sxyazi** [comment]
It's better for the Nushell wrapper to keep consistent with [the documentation](https://yazi-rs.github.io/docs/quick-start) - use `...` to simplify if-else blocks, add the `rm -p` parameter to permanently delete temporary files instead of sending them to the recycle bin:

```nu
def --env ya [...args] {
	let tmp = (mktemp -t "yazi-cwd.XXXXX")
	yazi ...$args --cwd-file $tmp
	let cwd = (open $tmp)
	if $cwd != "" and $cwd != $env.PWD {
		cd $cwd
	}
	rm -fp $tmp
}
```

File: tests/modules/programs/yazi/zsh-integration-enabled.nix [hunk 3]
```diff
@@ -4,7 +4,7 @@ let
   shellIntegration = ''
     function ya() {
       tmp="$(mktemp -t "yazi-cwd.XXXXX")"
```
Comment (sxyazi) [inline tests/modules/programs/yazi/zsh-integration-enabled.nix:6]:
```suggestion
      local tmp="$(mktemp -t "yazi-cwd.XXXXX")"
```

**Comment by XYenon** [comment]
@sxyazi I think we should make these changes on master and then backport them together?

**Comment by sxyazi** [comment]
Ah yep, that makes sense.

**Comment by musjj** [comment]
I think it would be better if you could specify a custom alias name. Maybe something like:
```nix
shellAlias = mkOption {
  type = types.str;
  default = "ya";
  description = "The alias name used when shell integration is enabled.";
};
```
Not sure if there's a more established pattern/convention though.

**Review by musjj — COMMENTED** [review]
Not super critical, but it would be nice if these functions can trigger shell completion. On my `zsh` config, this is how I did it:
```zsh
compdef _yazi ya
```
But I'm not really sure how you'd achieve it in other shells.

**Comment by XYenon** [comment]
@musjj I think we should backport these commits as-is first, to fix some bugs like https://github.com/nix-community/home-manager/issues/4822. And then add new features to the master branch.

**Comment by rycee** [comment]
Thanks! Merged to master now 🙂 I definitely agree that additional improvements should go through master first.

**Pull request merged**